### PR TITLE
Fix issue #90

### PR DESCRIPTION
**Pull request by ililikakis**
Adapt callback function has been extended by the input is_family so that num_elements can be set independently.
See issue #90 

**_All these boxes must be checked by the reviewers before merging the pull request:_**

- [x] The author added a BSD statement to `doc/` (or already has one)
- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request indroduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [x] All tests pass (in various configurations, this should be executed automatically in a github action)
- [x] New source/header files are properly added to the Makefiles
- [x] The reviewer executed the new code features at least once and checked the results manually
- [x] The code is covered in an existing or new test case
- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [x] Testing of this template: If you feel something is missing from this list, contact the developers

### CODE REVIEW

**Comment by Davknapp** [comment]
Just had a first look at it: You changed the BSD file for Niklas Böing. This file should not be touched in this PR. Please remove it from this PR. If there are errors in the file, please reach out to Niklas Böing to change it.

**Comment by ililikakis** [comment]
> Just had a first look at it: You changed the BSD file for Niklas Böing. This file should not be touched in this PR. Please remove it from this PR. If there are errors in the file, please reach out to Niklas Böing to change it.
It's done.

File: example/tutorials/t8_step3_adapt_forest.cxx [hunk 1]
```diff
@@ -84,21 +84,24 @@ struct t8_step3_adapt_data
...
- * \param [in] lelement_id The tree local index of the current element (or the first of the family).
- * \param [in] ts      The refinement scheme for this tree's element class.
- * \param [in] num_elements The number of elements. If this is > 1 we know that we look at a family.
- * \param [in] elements The element or family of elements to consider for refinement/coarsening.
+ * \param [in] forest       The current forest that is in construction.
+ * \param [in] forest_from  The forest from which we adapt the current forest (in our case, the uniform forest)
+ * \param [in] which_tree   The process local id of the current tree.
+ * \param [in] lelement_id  The tree local index of the current element (or the first of the family).
+ * \param [in] ts           The refinement scheme for this tree's element class.
+ * \param [in] is_family    If 1 considre family, if 0 do not.
```
Comment (Davknapp) [inline example/tutorials/t8_step3_adapt_forest.cxx:92]:
```suggestion
 * \param [in] is_family    if 1, the first \a num_elements entries in \a elements form a family. If 0, they do not.
```

File: src/t8_forest.h [hunk 2]
```diff
@@ -95,20 +95,21 @@ typedef void        (*t8_forest_replace_t) (t8_forest_t forest_old,
...
- *                         pointer to one element.
+ * Otherwise \a is_family must equal zero and we consider the first entry
+ * of the element array for refinement. 
+ * Entries of the element array beyond the first \a num_elements are undefined.
+ * \param [in] forest       the forest to which the new elements belong
+ * \param [in] forest_from  the forest that is adapted.
+ * \param [in] which_tree   the local tree containing \a elements
+ * \param [in] lelement_id  the local element id in \a forest_old in the tree of the current element
+ * \param [in] ts           the eclass scheme of the tree
+ * \param [in] is_family    if 1 considre family, if 0 do not.
```
Comment (Davknapp) [inline src/t8_forest.h:109]:
```suggestion
 * \param [in] is_family    if 1, the first \a num_elements entries in \a elements form a family. If 0, they do not.
```

File: src/t8_forest.h [hunk 3]
```diff
@@ -120,6 +121,7 @@ typedef int         (*t8_forest_adapt_t) (t8_forest_t forest,
                                           t8_locidx_t which_tree,
                                           t8_locidx_t lelement_id,
                                           t8_eclass_scheme_c * ts,
+                                          int is_family,
```
Comment (Davknapp) [inline src/t8_forest.h:124]:
```suggestion
                                          const int is_family,
```
is_family should not be changed during the adapt-callback, hence it should be const

**Review by Davknapp — CHANGES_REQUESTED** [review]

File: src/t8_forest.h (hunk 3 above)
Comment (ililikakis) [reply src/t8_forest.h:124]: That's right! But isn't that also the case for num_elements, so both should be const?

Comment (Davknapp) [reply src/t8_forest.h:124]: Yes, you are right! Can you change it too?

Comment (ililikakis) [reply src/t8_forest.h]: done

File: example/timings/time_new_refine.c [hunk 4]
```diff
@@ -34,13 +34,13 @@
 /* This function refines every element */
 static int
 t8_basic_adapt_refine (t8_forest_t forest, t8_locidx_t which_tree,
-                       t8_eclass_scheme_t * ts,
+                       t8_eclass_scheme_t * ts, int is_family,
```
Comment (Davknapp) [inline example/timings/time_new_refine.c:37]:
```suggestion
                       t8_eclass_scheme_t * ts, const int is_family,
```
```suggestion
                       t8_eclass_scheme_t * ts, int is_family,
```

File: example/timings/time_new_refine.c [hunk 5]
```diff
@@ -57,10 +57,10 @@ t8_basic_adapt_refine (t8_forest_t forest, t8_locidx_t which_tree,
 /* This function coarsens each element */
 static int
 t8_basic_adapt_coarsen (t8_forest_t forest, t8_locidx_t which_tree,
-                        t8_eclass_scheme_t * ts,
+                        t8_eclass_scheme_t * ts, int is_family,
```
Comment (Davknapp) [inline example/timings/time_new_refine.c:60]:
```suggestion
                        t8_eclass_scheme_t * ts, const int is_family,
```
```suggestion
                        t8_eclass_scheme_t * ts, int is_family,
```

Comment (Davknapp) [inline example/timings/time_new_refine.c:60]:
```suggestion
                        t8_eclass_scheme_t * ts, const int is_family,
```
```suggestion
                        t8_eclass_scheme_t * ts, int is_family,
```

File: example/timings/time_new_refine.c [hunk 6]
```diff
@@ -34,13 +34,13 @@
 /* This function refines every element */
 static int
 t8_basic_adapt_refine (t8_forest_t forest, t8_locidx_t which_tree,
-                       t8_eclass_scheme_t * ts,
+                       t8_eclass_scheme_t * ts, int is_family,
                        int num_elements, t8_element_t * elements[])
```
Comment (Davknapp) [inline example/timings/time_new_refine.c:38]:
```suggestion
                       const int num_elements, t8_element_t * elements[])
```
```suggestion
                       int num_elements, t8_element_t * elements[])
```

Comment (Davknapp) [inline example/timings/time_new_refine.c:38]:
```suggestion
                       const int num_elements, t8_element_t * elements[])
```
```suggestion
                       int num_elements, t8_element_t * elements[])
```

Comment (Davknapp) [inline example/timings/time_new_refine.c:38]:
```suggestion
                       const int num_elements, t8_element_t * elements[])
```
```suggestion
                       int num_elements, t8_element_t * elements[])
```

Comment (Davknapp) [inline example/timings/time_new_refine.c:38]:
```suggestion
                       const int num_elements, t8_element_t * elements[])
```
```suggestion
                       int num_elements, t8_element_t * elements[])
```

**Review by Davknapp — CHANGES_REQUESTED** [review]

**Pull request merged**